influxdb/src/main/scala/akka/stream/alpakka/influxdb/impl/InfluxDBFlowStage.scala

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.influxdb.impl
+
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+import akka.stream.alpakka.influxdb.{InfluxDBWriteMessage, InfluxDBWriteResult}
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import org.influxdb.InfluxDB
+import org.influxdb.impl.InfluxDBMapper
+
+import scala.collection.immutable
+import org.influxdb.BatchOptions
+import org.influxdb.dto.Point
+
+/**
+ * INTERNAL API
+ */
+private[influxdb] class InfluxDBFlowStage[T, C](
+    clazz: Option[Class[T]],
+    influxDB: InfluxDB
+) extends GraphStage[FlowShape[immutable.Seq[InfluxDBWriteMessage[T, C]], immutable.Seq[InfluxDBWriteResult[T, C]]]] {
+  private val in = Inlet[immutable.Seq[InfluxDBWriteMessage[T, C]]]("in")
+  private val out = Outlet[immutable.Seq[InfluxDBWriteResult[T, C]]]("out")
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    clazz match {
+      case Some(c) => new InfluxDBMapperRecordLogic
+      case None => new InfluxDBRecordLogic
+    }
+
+  sealed abstract class InfluxDBLogic extends GraphStageLogic(shape) with InHandler with OutHandler {
+
+    protected def write(messages: immutable.Seq[InfluxDBWriteMessage[T, C]]): Unit
+
+    setHandlers(in, out, this)
+
+    override def onPull(): Unit = if (!isClosed(in) && !hasBeenPulled(in)) pull(in)
+
+    override def onPush(): Unit = {
+      val messages = grab(in)
+      if (messages.nonEmpty) {
+
+        influxDB.enableBatch(BatchOptions.DEFAULTS)
+        write(messages)
+        val writtenMessages = messages.map(m => new InfluxDBWriteResult(m, Option.empty))
+        influxDB.close()
```

Comment:
Is it inefficient to call "close" here? Can we re-use the thread pool between elements?

Reply:
Batching can be enabled once. Therefore if two threads try to enable batch simultaneously an exception shall be thrown. Instead you enable batch once and then all write requests (originating from various threads) go through a buffer (link blocking queue) and the messages are flush through an interval  (default 1000 milliseconds) or when close is called.

Reply:
@adkafka

With regards to batching the reason for adding it, was to sent all the data at once instead of making write requests to InfluxDB one by one. Therefore the batch of writes on the Alpakka side would be processed as a batch on the client side too. Thus I am not actually using the batching functionality for the long term, instead I am using it in order to sent all the writes at once. 

There are various options with regards to the driver’s batching functionality.

1) Don’t add any batching through the driver at all. This will lead to writes being executed one by one.

2) Let the user handle the batching. Instead on enabling or closing the batching client side, the user will do so. He can use the same or various InfluxDB instances on the streams implemented and the user is going to decide when batching shall be enabled/disabled.

This might have some side effects. When batching is enabled (until the close operation is called). The data are going to be put in a queue and they are going to be processed either once a scheduler is called or when close is called. This shall lead to write requests from different flows (onPush) to be actually placed on a queue and get processed by a thread not managed by the Akka environment.

3) Another option is to enable batching per onPush calls (current solution). Instead of passing the InfluxDB instance, the flow shall create an instance from the configuration passed (just like the elastic and oriented implementation where the client is created on the spot) to execute the batches once off and then close.

Facilitating all of the options might make it bloated. Which one do you think is better (or if there is a better proposal)?

Reply:
The latest commit has removed any driver thread-based batching implementations and sends a batch of measurements using and http request. Since it is a blocking operation the logic shall be moved on the IODispatcher. 
Therefore the streams environment shall have full control over the request and the batch size.